golos.publication/golos.publication.cpp

```diff
@@ -511,101 +512,124 @@ void publication::send_deletevotes_trx(int64_t message_id, name author) {
-    require_auth(_self);
-    
-    tables::permlink_table permlink_table(_self, message_id.author.value);
-    auto permlink_index = permlink_table.get_index<"byvalue"_n>();
-    auto permlink_itr = permlink_index.find(message_id.permlink);
-    // rare case - not critical for performance
-    eosio::check(permlink_itr != permlink_index.end(), "Permlink doesn't exist.");
+void publication::send_closemssgs_trx() {
+    auto cur_time = static_cast<uint64_t>(eosio::current_time_point().time_since_epoch().count());
+    transaction trx(eosio::current_time_point() + eosio::seconds(config::closemssgs_expiration_sec));
+    trx.actions.emplace_back(action{permission_level(_self, config::code_name), _self, "closemssgs"_n, ""});
+    trx.delay_sec = 0;
+    trx.send(static_cast<uint128_t>(cur_time) << 64, _self, true);
+}
 
-    tables::message_table message_table(_self, message_id.author.value);
-    auto mssg_itr = message_table.find(permlink_itr->id);
-    eosio::check(mssg_itr != message_table.end(), "Message doesn't exist in cashout window.");
-    eosio::check(!mssg_itr->closed, "Message is closed.");
+void publication::close_messages() {
+    auto cur_time = static_cast<uint64_t>(eosio::current_time_point().time_since_epoch().count());
 
-    tables::reward_pools pools(_self, _self.value);
-    auto pool = get_pool(pools, mssg_itr->pool_date);
+    tables::reward_pools pools_table(_self, _self.value);
+    std::map<uint64_t, structures::rewardpool> pools;
 
-    eosio::check(pool->state.msgs != 0, "LOGIC ERROR! publication::payrewards: pool.msgs is equal to zero");
-    atmsp::machine<fixp_t> machine;
-    fixp_t sharesfn = set_and_run(
-            machine, 
-            pool->rules.mainfunc.code, 
-            {FP(mssg_itr->state.netshares)}, 
-            {{fixp_t(0), FP(pool->rules.mainfunc.maxarg)}});
-
-    asset mssg_reward;
-    auto state = pool->state;
-    int64_t payout = 0;
-    if(state.msgs == 1) {
-        payout = state.funds.amount; //if we have the only message in the pool, the author receives a reward anyway
-        eosio::check(state.rshares == mssg_itr->state.netshares,
-                "LOGIC ERROR! publication::payrewards: pool->rshares != mssg_itr->netshares for last message");
-        eosio::check(state.rsharesfn == sharesfn.data(),
-                "LOGIC ERROR! publication::payrewards: pool->rsharesfn != sharesfn.data() for last message");
-        state.funds.amount = 0;
-        state.rshares = 0;
-        state.rsharesfn = 0;
-    }
-    else {
-        auto total_rsharesfn = WP(state.rsharesfn);
+    size_t i = 0;
+    tables::message_table message_table(_self, _self.value);
+    auto message_index = message_table.get_index<"bycashout"_n>();
+    for (auto mssg_itr = message_index.begin(); mssg_itr != message_index.end(); ++mssg_itr) {
+        if (config::max_closed_posts_per_action <= i++) {
+            send_closemssgs_trx();
+            return;
```

Comment:
As I understand in this case it's need to execute cycle in 616 line.

Reply:
And better call `send_closemssgs_trx` if really need it (not all posts was processed).